src/v/pandaproxy/schema_registry/validation.cc

```diff
@@ -0,0 +1,411 @@
+#include "pandaproxy/schema_registry/seq_writer.h"
+#include "pandaproxy/schema_registry/sharded_store.h"
+#include "pandaproxy/schema_registry/subject_name_strategy.h"
+#include "pandaproxy/schema_registry/types.h"
+#include "storage/parser_utils.h"
+#include "utils/vint.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/sstring.hh>
+#include <seastar/coroutine/exception.hh>
+
+#include <absl/algorithm/container.h>
+
+#include <iterator>
+#include <optional>
+#include <stdexcept>
+#include <type_traits>
+
+namespace pandaproxy::schema_registry {
+
+namespace {
+
+using field = schema_id_cache::field;
+
+auto to_string_view(field f) {
+    switch (f) {
+    case field::key:
+        return "key";
+    case field::val:
+        return "value";
+    }
+};
+
+auto make_subject(
+  subject_name_strategy sns,
+  const model::topic& t,
+  field f,
+  std::string_view record_name) {
+    switch (sns) {
+    case subject_name_strategy::topic_name:
+        return subject{ssx::sformat("{}-{}", t(), to_string_view(f))};
+    case subject_name_strategy::record_name:
+        return subject{record_name};
+    case subject_name_strategy::topic_record_name:
+        return subject{ssx::sformat("{}-{}", t(), record_name)};
+    }
+};
+
+std::vector<int32_t> get_proto_offsets(iobuf_const_parser& p) {
+    // The encoding is a length, followed by indexes into the file or message.
+    // Each number is a zigzag encoded integer.
+    std::vector<int32_t> offsets;
+    auto [result, bytes_read] = p.read_varlong();
+    if (!bytes_read) {
+        return {};
+    }
+    offsets.resize(result);
+    for (auto& o : offsets) {
+        std::tie(o, bytes_read) = p.read_varlong();
+        if (!bytes_read) {
+            return {};
+        }
+    }
+    if (offsets.empty()) {
+        // If the decoded length is 0, then assume the optimised encoding for
+        // using the first message in the file
+        offsets.push_back(0);
+    }
+    return offsets;
+}
+
+ss::future<std::optional<ss::sstring>> get_record_name(
+  pandaproxy::schema_registry::sharded_store& store,
+  field field,
+  const model::topic& topic,
+  subject_name_strategy sns,
+  const canonical_schema_definition& schema,
+  std::optional<std::vector<int32_t>>& offsets) {
+    if (sns == subject_name_strategy::topic_name) {
+        // Result is succesfully nothing
+        co_return "";
+    }
+
+    switch (schema.type()) {
+    case schema_type::avro: {
+        auto s = co_await make_avro_schema_definition(
+          store, {subject("r"), {schema.raw(), schema.type()}});
+        co_return s().root()->name().fullname();
+    } break;
+    case schema_type::protobuf: {
+        if (!offsets) {
+            co_return std::nullopt;
+        }
+        auto s = co_await make_protobuf_schema_definition(
+          store, {subject("r"), {schema.raw(), schema.type()}});
+        auto r = s.name(*offsets);
```

Comment:
```
protobuf_schema_definition::name(std::vector<int> fields) 
```
takes the arg by value, so maybe either `s.name(std::move(*offsets))`, or make params to both functions const reference?

Reply:
Done